File: contribution-pipeline/scheduler/backend-py/src/scheduler/tenants_create.py
"""
CLI: 创建租户

用法:
    python -m scheduler.tenants_create --slug=acme --name="Acme Corp"
    python -m scheduler.tenants_create --slug=acme --name="Acme Corp" --plan=enterprise

slug 只能含小写字母 / 数字 / 连字符, 长度 2~32 (路由 / 日志 / 子域名复用此值)。
创建后输出 tenant id, 后续 keys_create 时用 --tenant=<slug>。
"""
from __future__ import annotations
import asyncio
import re
import sys

from dotenv import load_dotenv

from .db import init_pool, close_pool, as_system

SLUG_RE = re.compile(r"^[a-z0-9](?:[a-z0-9-]{0,30}[a-z0-9])?$")


def parse_args() -> tuple[str, str, str]:
    args: dict[str, str] = {}
    for a in sys.argv[1:]:
        m = re.match(r"^--([^=]+)=(.*)$", a)
        if m:
            args[m.group(1)] = m.group(2)
    if not args.get("slug") or not args.get("name"):
        print("用法: python -m scheduler.tenants_create --slug=<slug> --name=<name> [--plan=standard]", file=sys.stderr)
        sys.exit(1)
    if not SLUG_RE.match(args["slug"]):
        print(f"slug 不合法: {args['slug']}", file=sys.stderr)
        print("规则: 小写字母 / 数字 / 连字符, 长度 2~32, 不能以连字符开头或结尾", file=sys.stderr)
        sys.exit(1)
    return args["slug"], args["name"], args.get("plan", "standard")


async def _amain() -> None:
    slug, name, plan = parse_args()
    await init_pool()
    try:
        async def _do(conn):
            return await conn.fetchrow(
                "INSERT INTO tenants (slug, name, plan) VALUES ($1, $2, $3) RETURNING id",
                slug, name, plan,
            )
        row = await as_system(_do)
        print("✓ 租户已创建")
        print(f"  id:    {row['id']}")
        print(f"  slug:  {slug}")
        print(f"  name:  {name}")
        print(f"  plan:  {plan}")
        print()
        print(f"下一步: python -m scheduler.keys_create --name=<service> --tenant={slug}")
    except Exception as e:
        msg = str(e)
        if "duplicate key" in msg:
            print("✗ slug 已存在,换一个", file=sys.stderr)
        else:
            print(e, file=sys.stderr)
        sys.exit(1)
    finally:
        await close_pool()


def main() -> None:
    load_dotenv()
    asyncio.run(_amain())


if __name__ == "__main__":
    main()

File: contribution-pipeline/scheduler/backend/src/tenants-create.ts
/**
 * CLI: 创建租户
 *
 * 用法:
 *   npm run tenants:create -- --slug=acme --name="Acme Corp"
 *   npm run tenants:create -- --slug=acme --name="Acme Corp" --plan=enterprise
 *
 * slug 只能含小写字母 / 数字 / 连字符, 长度 2~32 (路由 / 日志 / 子域名复用此值)。
 * 创建后输出 tenant id, 后续 keys:create 时用 --tenant=<slug>。
 */

import { sql } from "./db.ts";

const SLUG_RE = /^[a-z0-9](?:[a-z0-9-]{0,30}[a-z0-9])?$/;

function parseArgs(): { slug: string; name: string; plan: string } {
  const args: Record<string, string> = {};
  for (const a of process.argv.slice(2)) {
    const m = a.match(/^--([^=]+)=(.*)$/);
    if (m) args[m[1]] = m[2];
  }
  if (!args.slug || !args.name) {
    console.error("用法: npm run tenants:create -- --slug=<slug> --name=<name> [--plan=standard]");
    process.exit(1);
  }
  if (!SLUG_RE.test(args.slug)) {
    console.error(`slug 不合法: ${args.slug}`);
    console.error("规则: 小写字母 / 数字 / 连字符, 长度 2~32, 不能以连字符开头或结尾");
    process.exit(1);
  }
  return { slug: args.slug, name: args.name, plan: args.plan ?? "standard" };
}

async function main() {
  const { slug, name, plan } = parseArgs();
  const rows = await sql<{ id: string }[]>`
    INSERT INTO tenants (slug, name, plan)
    VALUES (${slug}, ${name}, ${plan})
    RETURNING id
  `;
  console.log("✓ 租户已创建");
  console.log(`  id:    ${rows[0].id}`);
  console.log(`  slug:  ${slug}`);
  console.log(`  name:  ${name}`);
  console.log(`  plan:  ${plan}`);
  console.log("");
  console.log(`下一步: npm run keys:create -- --name=<service> --tenant=${slug}`);
  await sql.end();
}

main().catch((e) => {
  if (typeof e?.message === "string" && /duplicate key/.test(e.message)) {
    console.error("✗ slug 已存在,换一个");
  } else {
    console.error(e);
  }
  process.exit(1);
});
